Minor changes for compatibility with Mingw-w64 in Windows 64-bit.

guy-rouillier opened the pull request:
I've tested these changes generating a 64-bit and 32-bit binary on Windows 7 64-bit using Mingw-w64 on msys64.  I also tested  generating a 64-bit and 32-bit binary on Ubuntu Mate 21.10 64-bit.  I used the tvedit example for testing, using the makefiles for building.  For the 32-bit version, the -m32 flag must be specified for CFLAGS and CXXFLAGS during configure.  All 4 versions worked as expected.

After some effort, I've now got it building and running with MSVC, both in 32-bit and 64-bit mode.  I had to make some small changes to get this to work.  Let me know if you'd like to incorporate those changes.

I'm not going to bother with the Borland compiler.  I actually have the 5.5 compiler, with all the original printed manuals! But at this point, that compiler is over 25 years old, and in my opinion does not warrant putting in the effort.  It doesn't support any C++ standard in the last 20 years.

guy-rouillier commented on tvision/classes/wingr/wingrscr.cc:
```diff
@@ -734,8 +734,13 @@ int TScreenWinGr::System( const char * command
-  }
+  {
+	#if defined(TVCompf_MinGW) && defined(HAVE_64BITS_POINTERS)
+		return( *pidChild= (uint64_t)pi.hProcess );  // Give process identifier
```
Without this cast, MingW will not compile this code in 64-bit mode. I don't want to require -fpermissive on the whole project.

```
g++  -I../include        -O2 -Wall -ggdb3  -c ../classes/wingr/wingrscr.cc -o obj/wingrscr.o  

../classes/wingr/wingrscr.cc: In static member function 'static int TScreenWinGr::System(const char*, unsigned int*, int, int, int)':  

../classes/wingr/wingrscr.cc:740:36: error: cast from 'HANDLE' {aka 'void*'} to 'unsigned int' loses precision [-fpermissive]  
  740 |                 return( *pidChild= (pid_t)pi.hProcess );  // Give process identifier  

      |                                    ^~~~~~~~~~~~~~~~~~  

make[1]: *** [librhtv.umk:1441: obj/wingrscr.o] Error 1 
```

set-soft replied:
But with this cast you don't solve the problem. You are just forcing the compilation, if anybody tries to use this value things won't work because you are discarding 32 bits of the 64 bits.

guy-rouillier replied:
Please see latest commit, I believe I've addressed this issue.

guy-rouillier commented on tvision/include/compatlayer.h:
```diff
@@ -2108,7 +2109,9 @@ CLY_CFunc int  CLY_getcurdir(int drive, char *buffer);
-  typedef long ssize_t;
+	#ifndef _SSIZE_T_DEFINED
+		typedef long ssize_t;
+	#endif
```
Salvador, I tried to follow your suggestion on the previous pull request to avoid defining Uses_CLY_ssize_t at line 560.  However, <sys/types.h> is only included immediately prior to here, around line 2108, so _SSIZE_T_DEFINED is not defined up at line 560.  I didn't want to start moving things around, since there are so many includes and dependencies between them.

guy-rouillier replied:
I just checked out the source for mingw_w64 from here: https://github.com/mingw-w64/mingw-w64

Going all the way back to version 1.x in 2009, mingw_w64 defined ssize_t. The older, original mingw has moved here: https://osdn.net/projects/mingw/.  It is still only a 32-bit compiler.  I don't think we should support it.  What is your opinion?  I have not checked this older original mingw to see if/when it first added ssize_t.

I will try moving the include of sys/types.h up above Uses_CLY_ssize_t at line 560.  If it doesn't cause too many problems on all the compilers we want to support, then I suppose that is the better solution.

set-soft replied:
Why not just use a define like I suggest? 
About the old MinGW: which version was this? 
You should look for some definition that allows differentiating these versions. With this we can just avoid the Uses_CLY_ssize_t for modern MinGW versions.

guy-rouillier replied:
The original MinGW is here: https://sourceforge.net/projects/mingw/.  You'll see there that it has moved to a new location. MingW-w64 was forked from the original years ago, and has moved on, while the original project has not seen significant updates, though it continues to generate releases.  The original is still only 32-bit.  My opinion is we should test for the newer MingW-w64, and if it is not being used, abort the build and tell the user to use MingW-w64.  Do you have a particular reason for wanting to support the older 32-bit version?  I don't think many people are still using it.  Plus, we have a big enough challenge supporting just MingW-w64, along with other compilers (MSC, GCC).

guy-rouillier replied:
Please see latest commit, I believe I've addressed this issue.

guy-rouillier commented on tvision/include/tv/tvconfig.h:
```diff
@@ -33,7 +33,7 @@ const int maxReplaceStrLen = 80;
 #ifndef pid_t
-# define pid_t int
+# define pid_t unsigned int
 #endif
```
Windows defines a PID to be a DWORD, which is an unsigned int, On Linux a PID only uses 22 bits, so an unsigned int should be adequate for both.

guy-rouillier replied:
I did a bit of research.  Turns out, for compatibility between 32-bit and 64-bit processes, only the lower 32-bits are used in hProcess.  But I have an alternative I'm working on, originating in your observation that hProcess is not used anywhere else, which I confirmed by searching the code.  In the call to CreateProcess, one of the return values is the PID in pi.dwProcessId, which is a DWORD, a 32-bit value.  So, rather than store hProcess in pidChild, we can store the actual PID instead.

guy-rouillier replied:
Please see latest commit, I believe I've addressed this issue.

set-soft reviewed (commented):
Please see the notes

guy-rouillier commented:
Please see latest commit, I believe I've addressed the issues you've identified.  Thanks.

set-soft commented:
Hi @guy-rouillier !
I incorporated most changes. We need to solve some details:

1. Why is pid_t declared as `unsigned int` instead of `int`?
2. We must rename svrapi.h to something else and use it only when needed. We don't know if the svrapi.h content is good for old compilers.

guy-rouillier commented:
1. pid_t on both Linux and Windows is non-negative only.  But I just did 
some research, and see that POSIX defines it as a signed number.  So, 
revert to int.  Good catch.  The change I made in wingr\wingrscr.cc to 
assign dwProcessId instead of hProcess to the pid variable changes the 
returned value to a DWORD, which is unsigned.  But as I mentioned, 
Windows uses less than the full 32-bits, so stuffing that value into a 
signed variable should not cause issues.

2. svrapi.h was defined in 1995.  I doubt we'll be encountering any 
compilers older than that.  This include is only used in one source 
file, tvision\compat\unc_95.c.  The header is protected by a sentinel 
"#ifndef SVRAPI_INCLUDED", so if the build finds this file somewhere 
else, this local version will be skipped.  I don't see any harm in 
keeping it as is.  As I mentioned in my last commit, the only reason I 
added it was that this on source file needs it, and current Windows 
toolkits no longer include it.

A bit more on that. The reason new toolkits don't include svrapi.h is 
that it is obsolete.  Those functions are now defined in LMaccess.h.  
The LM*.h headers have been ported to MinGW and GCC.  So, we should put 
it on our list to get rid of svrapi.h entirely, and switch to 
LMaccess.h.  But I don't know the availability of the LM*.h headers 
under DOS, and can't put the time in to figure that out right now.

I just did local builds of MinGW and MSC, both 32 and 64-bit, after 
changing pid_t back to int, and all four builds succeeded (and tvedit 
ran without issue.) So, you may change that back.

Thanks.

--
Guy Rouillier

On 5/24/2022 7:48:11 AM, "Salvador E. Tropea" ***@***.***> 
wrote:

>
>Hi @guy-rouillier <https://github.com/guy-rouillier> !
>I incorporated most changes. We need to solve some details:
>
>Why is pid_t declared as unsigned int instead of int?
>We must rename svrapi.h to something else and use it only when needed. 
>We don't know if the svrapi.h content is good for old compilers.
>—
>Reply to this email directly, view it on GitHub 
><https://github.com/set-soft/tvision/pull/4#issuecomment-1135813087>, 
>or unsubscribe 
><https://github.com/notifications/unsubscribe-auth/AA3JYPNZZEQECGQWQ2XJTJDVLS6XXANCNFSM5RKESD4A>.
>You are receiving this because you were mentioned.Message ID: 
***@***.***>
>
